certbot/interfaces.py

```diff
@@ -238,6 +238,12 @@ class IConfig(zope.interface.Interface):
         "This only affects the port Certbot listens on. "
         "A conforming ACME server will still attempt to connect on port 80.")
 
+    http01_use_tls = zope.interface.Attribute(
+        "Runs the standalone server with a self signed certificate. "
+        "Useful for when you need to validate an http-01 challenge with "
+        "an https endpoint or if you have an upstream proxy redirecting "
+        "all http traffic to https.")
```

Comment:
I'm not sure I understand this correctly. If you have a proxy that redirects http/80 to https/443 (via HTTP 301/302), you're good. The upstream HTTPS server may indeed operate as a HTTPS server, that's between you and the proxy. I'll try to write this as a sequence of requests:

1. ACME client asks ACME server to validate domain
2. ACME server connects to your (reverse) proxy on port 80, requesting `http://example.com/.well-known/acme-challenge/{token}`
3. Reverse proxy responds with HTTP 301 to `https://example.com/.well-known/acme-challenge/{token}`
4. ACME server connects to reverse proxy on on port 443 via TLS, requesting `https://example.com/.well-known/acme-challenge/{token}`
5. Reverse proxy connects to upstream/backend server, requesting `/.well-known/acme-challenge/{token}`. This can be either HTTP or HTTPS, on any port - the ACME server wouldn't known or care.
6. Upstream server responds with challenge token, reverse proxy forwards that response, challenge passes, all is well.

This is how e.g. people use Let's Encrypt behind CloudFlare today. Is this what you were referring to, or are we talking about different things? 😇